controllers/plan_controller.go

```diff
@@ -0,0 +1,147 @@
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	response.Success("Successfully retrieved plans", page.Page, totalData, plans)
+	return c.JSON(http.StatusOK, response)
+}
+
+// Get plan by ID
+func GetPlanController(c echo.Context) error {
+	var response models.GeneralResponse
+	var err models.CustomError
+
+	var readablePlan models.ReadablePlan
+	var planObject models.Plan
+
+	// Extract the plan ID from the URL parameter
+	planID := c.Param("id")
+
+	// Convert the plan ID string to int
+	planObject.InsertID(planID, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	// Retrieve the plan from the database
+	database.GetPlan(&planObject, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	// Convert the plan object to a readable format
+	readablePlan.ToReadablePlan(&planObject)
+
+	response.Success(http.StatusOK, "Successfully retrieved plan", readablePlan)
+	return c.JSON(http.StatusOK, response)
+}
+
+// Create new plan
+func CreatePlanController(c echo.Context) error {
+	var response models.GeneralResponse
+	var err models.CustomError
+
+	var planObject models.Plan
+
+	err.ErrorMessage = c.Bind(&planObject)
+	if err.IsError() {
+		err.StatusCode = http.StatusBadRequest
+		err.ErrorReason = "Invalid request body"
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	database.CreatePlan(&planObject, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	response.Success(http.StatusCreated, "Successfully created a new plan", planObject)
+	return c.JSON(response.StatusCode, response)
+}
+
+// Update plan
+func UpdatePlanController(c echo.Context) error {
+	var response models.GeneralResponse
+	var err models.CustomError
+
+	var updatedPlan models.Plan
+
+	planID := c.Param("id")
+
+	err.ErrorMessage = c.Bind(&updatedPlan)
+	if err.IsError() {
+		err.StatusCode = http.StatusBadRequest
+		err.ErrorReason = "Invalid request body"
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	updatedPlan.InsertID(planID, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	database.UpdatePlan(&updatedPlan, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	response.Success(http.StatusOK, "Successfully updated plan", updatedPlan)
+	return c.JSON(response.StatusCode, response)
+}
+
+// Delete plan
+func DeletePlanController(c echo.Context) error {
+	var response models.GeneralResponse
+	var err models.CustomError
+
+	planID := c.Param("id")
+
+	var deletedPlan models.Plan
+	deletedPlan.InsertID(planID, &err)
+	if err.IsError() {
+		response.ErrorOcurred(&err)
+		return c.JSON(response.StatusCode, response)
+	}
+
+	database.DeletePlan(&deletedPlan, &err)
```

Comment:
data yang ingin dihapus dicari terlebih dahulu. sehingga ketika data yang ingin dihapus tidak ada (not exist) akan mengembalikan response record not found. contohnya dapat dilihat pada user_controller.go